test/abicheck.sh

```diff
@@ -0,0 +1,152 @@
+
+Obeys CHOST, CONFIGURE_ARGS, CFLAGS, and LDFLAGS.
+
+Requires libabigail (on Ubuntu, install package abigail-tools).
+_EOF_
+}
+
+# Print the multiarch tuple for the current (non-cross) machine,
+# or the empty string if unavailable.
+detect_chost() {
+    dpkg-architecture -qDEB_HOST_MULTIARCH ||
+     $CC -print-multiarch ||
+     $CC -print-search-dirs | sed 's/:/\n/g' | grep -E '^/lib/[^/]+$' | sed 's%.*/%%' ||
+     true
+}
+
+if ! test -f "configure"
+then
+  echo "Please run from top of source tree"
+  exit 1
+fi
+
+suffix="-ng"
+CONFIGURE_ARGS_NG="$CONFIGURE_ARGS"
+refresh=false
+for arg
+do
+  case "$arg" in
+  --zlib-compat)
+    suffix=""
+    CONFIGURE_ARGS_NG="$CONFIGURE_ARGS_NG --zlib-compat"
+    ;;
+  --refresh)
+    refresh=true
+    ;;
+  --help)
+    usage
+    exit 0
+    ;;
+  *)
+    echo "Unknown arg '$arg'"
+    usage
+    exit 1
+    ;;
+  esac
+done
+
+# Choose reference repo and commit
+if test "$suffix" = ""
+then
+  # Reference is zlib 1.2.11
+  ABI_GIT_REPO=https://github.com/madler/zlib.git
+  ABI_GIT_COMMIT=v1.2.11
+else
+  # Reference should be the tag for zlib-ng 2.0
+  # but until that bright, shining day, use some
+  # random recent SHA.  Annoyingly, can't shorten it.
+  ABI_GIT_REPO=https://github.com/zlib-ng/zlib-ng.git
+  ABI_GIT_COMMIT=7b4e7f158ada676db0d59ee98fde79d1968ebac8
+fi
+# FIXME: even when using a tag, check the hash.
+
+# Test compat build for ABI compatibility with zlib
+if test "$CHOST" = ""
+then
+  # Note: don't export CHOST here, as we don't want configure seeing it
+  # when it's just the name for the build machine.
+  # Leave it as a plain shell variable, not an environment variable.
+  CHOST=$(detect_chost)
+  # Support -m32 for non-cross builds.
+  case "$CFLAGS" in
+  *-m32*) M32="-m32";;
+  *) M32="";;
+  esac
+fi
+if test "$CHOST" = ""
+then
+  echo "pkgcheck-abi-check SKIP, as we don't know CHOST"
+  exit 0
+fi
+
+ABIFILE="test/abi/zlib$suffix-$ABI_GIT_COMMIT-$CHOST$M32.abi"
+
+abidw --version
+
+if $refresh
+then
+  # Check out reference source
+  rm -rf btmp1
+  mkdir -p btmp1/src.d
+  cd btmp1/src.d
+  git init
+  git remote add origin $ABI_GIT_REPO
+  git fetch origin $ABI_GIT_COMMIT
+  git reset --hard FETCH_HEAD
+  cd ..
+  # Build unstripped, uninstalled, very debug shared library
+  # Canonicalize CHOST to work around bug in original zlib
+  env CHOST=$(sh tools/config.sub $CHOST) CFLAGS="$CFLAGS -ggdb" src.d/configure $CONFIGURE_ARGS
+  make -j2
+  cd ..
+  # Find shared library, extract its abi
+  dylib1=$(find btmp1 -type f -name '*.dylib*' -print -o -type f -name '*.so.*' -print)
+  abidw $dylib1 > "$ABIFILE"
+  # Maintainers may wish to check $ABIFILE into git when a new
+  # target is added, or when a major release happens that is
+  # intended to change the ABI.  Alternately, this script could
+  # just always rebuild the reference source, and dispense with
+  # caching abi files in git (but that would slow builds down).
```

Comment:
Do you know how long this would add?

Reply:
@dankegel *nudge*

Reply:
@Dead2 perhaps we should move ahead with this PR.

Reply:
I have an updated branch with fixes, will try to update this Mr tomorrow.